File: apps/ios/Packages/StoryArcKit/Sources/LibraryFeature/DownloadBanner.swift
public import SwiftUI

internal import DesignSystem
internal import StoryArcCore

/// What the queue is doing, along the foot of the catalogue.
///
/// One line for the download at the front and a count for the rest, because a reader
/// browsing a catalogue wants to keep browsing — a list of six transfers belongs in
/// Settings, not over the grid they are reading.
///
/// `offline-downloads` wants "progress ... visible on the publication and in a single
/// downloads view". This is the first half; Settings › Downloads and storage is the second.
struct DownloadBanner: View {
    @Environment(\.theme) private var theme

    let download: Download
    /// How many more are behind this one.
    let others: Int
    let onCancel: () -> Void
    let onResume: () -> Void

    var body: some View {
        HStack(spacing: StoryArcSpace.md) {
            VStack(alignment: .leading, spacing: StoryArcSpace.hair) {
                Text(title)
                    .textRole(.footnote)
                    .foregroundStyle(theme.palette.textPrimary)
                    .lineLimit(1)

                if others > 0 {
                    Text("downloads.queued \(others)", bundle: .module)
                        .textRole(.caption)
                        .foregroundStyle(theme.palette.textSecondary)
                }
            }

            Spacer(minLength: 0)

            action
        }
        .padding(.horizontal, StoryArcSpace.gutter)
        .padding(.vertical, StoryArcSpace.sm)
        .frame(maxWidth: .infinity)
        .background(.bar)
    }

    private var title: String {
        switch download.state {
        case let .failed(reason, _):
            reason
        case .paused:
            String(
                format: String(localized: "downloads.pausedTitle", bundle: .module),
                download.title
            )
        case .queued, .running, .finished:
            String(
                format: String(localized: "catalogue.acquire.fetching", bundle: .module),
                download.title
            )
        }
    }

    /// One button, and which one depends on what would help.
    ///
    /// A failed or paused download offers a retry; one that is moving offers a stop. Two
    /// buttons on a strip this size is a strip nobody can hit either half of.
    @ViewBuilder
    private var action: some View {
        switch download.state {
        case .failed, .paused:
            Button(action: onResume) {
                Text("downloads.retry", bundle: .module)
            }
            .buttonStyle(.bordered)
            .controlSize(.small)
        case .queued, .running, .finished:
            Button(role: .destructive, action: onCancel) {
                Text("downloads.stop", bundle: .module)
            }
            .buttonStyle(.bordered)
            .controlSize(.small)
        }
    }
}

File: apps/android/feature/library/src/main/kotlin/app/storyarc/feature/library/DownloadBanner.kt
package app.storyarc.feature.library

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.pluralStringResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextOverflow
import app.storyarc.core.designsystem.theme.LocalStoryArcPalette
import app.storyarc.core.designsystem.tokens.StoryArcSpace
import app.storyarc.core.model.Download

/**
 * What the queue is doing, along the foot of the screen that started it.
 *
 * One line for the download at the front and a count for the rest, because a reader browsing
 * a catalogue wants to keep browsing -- a list of six transfers belongs in Settings, not over
 * the grid they are reading. Shared by the browser and the publication detail screen: the
 * download can be started from either, and it has to be visible on the one it was started
 * from. iOS's `DownloadBanner` is the same strip.
 */
@Composable
internal fun DownloadBanner(
    download: Download,
    others: Int,
    onCancel: () -> Unit,
    onResume: () -> Unit,
) {
    val palette = LocalStoryArcPalette.current
    val failed = download.state as? Download.State.Failed
    val paused = download.state as? Download.State.Paused

    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(StoryArcSpace.sm),
        modifier = Modifier
            .fillMaxWidth()
            .background(palette.surfaceRaised)
            .padding(horizontal = StoryArcSpace.gutter, vertical = StoryArcSpace.sm),
    ) {
        Column(modifier = Modifier.weight(1f)) {
            Text(
                text = when {
                    failed != null -> failed.reason
                    paused != null -> stringResource(R.string.downloads_paused_title, download.title)
                    else -> stringResource(R.string.catalogue_acquire_fetching, download.title)
                },
                style = MaterialTheme.typography.bodySmall,
                color = palette.textPrimary,
                maxLines = 2,
                overflow = TextOverflow.Ellipsis,
            )
            if (others > 0) {
                Text(
                    text = pluralStringResource(R.plurals.downloads_queued, others, others),
                    style = MaterialTheme.typography.labelSmall,
                    color = palette.textSecondary,
                )
            }
        }

        // One button, and which one depends on what would help. Two on a strip this size is a
        // strip nobody can hit either half of.
        if (failed != null || paused != null) {
            TextButton(onClick = onResume) { Text(stringResource(R.string.downloads_retry)) }
        } else {
            TextButton(onClick = onCancel) { Text(stringResource(R.string.downloads_stop)) }
        }
    }
}
